sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/UnpredictableHashSaltCheck.kt

```diff
@@ -0,0 +1,150 @@
+    FunMatcher(qualifier = "java.lang.String", name = "getBytes")
+)
+private val BYTE_ARRAY_CONSTRUCTOR = ConstructorMatcher("kotlin.ByteArray")
+private val BYTE_ARRAY_CONSTRUCTOR_SIZE_ARG_ONLY = ConstructorMatcher("kotlin.ByteArray") { withArguments("kotlin.Int") }
+private val SECURE_RANDOM_FUNS = FunMatcher(qualifier = "java.security.SecureRandom")
+
+private const val MSG_MAKE_UNPREDICTABLE = "Make this salt unpredictable."
+private const val MSG_ADD_SALT = "Add an unpredictable salt value to this hash."
+private const val MSG_MIN_LEN = "Make this salt at least 16 bytes."
+private const val SMSG_PREDICTABLE_SALT = "Predictable salt value"
+
+@Rule(key = "S2053")
+class UnpredictableHashSaltCheck : CallAbstractCheckWithData<Int>() {
+    override val functionsToVisit = mapOf(
+        ConstructorMatcher("$SPECS_PACKAGE.$KEY_SPEC_FUN_NAME") to 1,
+        ConstructorMatcher("$SPECS_PACKAGE.$PARAMETER_SPEC_FUN_NAME") to 0,
+    )
+
+    override fun visitFunctionCall(
+        callExpression: KtCallExpression,
+        resolvedCall: ResolvedCall<*>,
+        saltArgIndex: Int,
+        kotlinFileContext: KotlinFileContext
+    ) {
+        if (callExpression.valueArguments.size < 2) {
+            kotlinFileContext.reportIssue(callExpression, MSG_ADD_SALT)
+            return
+        }
+
+        val bindingContext = kotlinFileContext.bindingContext
+        val saltArg = resolvedCall.valueArgumentsByIndex?.elementAtOrNull(saltArgIndex)
+            ?.arguments?.elementAtOrNull(0)?.getArgumentExpression() ?: return
+
+        val predictedSaltValue = saltArg.predictRuntimeValueExpression(bindingContext) ?: return
+
+        if (predictedSaltValue.isInitializedFromString(bindingContext)) {
+            kotlinFileContext.reportIssue(
+                saltArg,
+                MSG_MAKE_UNPREDICTABLE,
+                listOf(kotlinFileContext.secondaryOf(predictedSaltValue, SMSG_PREDICTABLE_SALT))
+            )
+            return
+        }
+
+        val saltInitializer = predictedSaltValue.getResolvedCall(bindingContext)
+            ?.takeIf { it matches BYTE_ARRAY_CONSTRUCTOR } ?: return
+
+        if (saltInitializer.byteArrayInitSizeTooSmall(bindingContext)) {
+            kotlinFileContext.reportIssue(saltArg, MSG_MIN_LEN, listOf(kotlinFileContext.secondaryOf(predictedSaltValue)))
+        }
+
+        if (saltInitializer matches BYTE_ARRAY_CONSTRUCTOR_SIZE_ARG_ONLY && saltArg.isInitializedPredictably(bindingContext)) {
+            kotlinFileContext.reportIssue(
+                saltArg,
+                MSG_MAKE_UNPREDICTABLE,
+                listOf(kotlinFileContext.secondaryOf(predictedSaltValue, SMSG_PREDICTABLE_SALT))
+            )
+        }
+    }
+
+    /**
+     * Checks whether the expression is a call, matches the FunMatchers in [STRING_TO_BYTE_FUNS] and is called on a constant string value.
+     */
+    private fun KtExpression.isInitializedFromString(bindingContext: BindingContext) =
```

Comment:
It is and I want to reuse it for SONARKT-145. I left it in the check for now because it didn't seem generic enough to put into `ApiExtensions`, so I thought I'll create a shared super class for both checks and pull this up. WDYT?

Reply:
I think It's ok, to put it in Api extensions like this. I still prefer association over inheritance, where possible :)